### ISSUE TITLE
Stack trace when fail_function returns false and data is a dictionary

### ISSUE THREAD

**Issue by amendlik** [saltstack/salt#18806]
Running an orchestration with ```salt.function``` where that function returns a dictionary (like ```cmd.run_all```) and a ```fail_function``` that returns ```False``` results in a stack trace on the minion and the master.

```
[ERROR   ] An exception occurred in this state: Traceback (most recent call last):
  File "/source/salt/salt/state.py", line 1533, in call
    **cdata['kwargs'])
  File "/source/salt/salt/states/saltmod.py", line 418, in function
    __opts__,
  File "/source/salt/salt/output/__init__.py", line 121, in out_format
    return get_printout(out, opts)(data).rstrip()
  File "/source/salt/salt/output/highstate.py", line 74, in output
    return _format_host(host, hostdata)[0]
  File "/source/salt/salt/output/highstate.py", line 119, in _format_host
    key=lambda k: data[k].get('__run_num__', 0)):
  File "/source/salt/salt/output/highstate.py", line 119, in <lambda>
    key=lambda k: data[k].get('__run_num__', 0)):
AttributeError: 'int' object has no attribute 'get'

dev:
----------
          ID: failtest
    Function: salt.function
        Name: cmd.run_all
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/source/salt/salt/state.py", line 1533, in call
                  **cdata['kwargs'])
                File "/source/salt/salt/states/saltmod.py", line 418, in function
                  __opts__,
                File "/source/salt/salt/output/__init__.py", line 121, in out_format
                  return get_printout(out, opts)(data).rstrip()
                File "/source/salt/salt/output/highstate.py", line 74, in output
                  return _format_host(host, hostdata)[0]
                File "/source/salt/salt/output/highstate.py", line 119, in _format_host
                  key=lambda k: data[k].get('__run_num__', 0)):
                File "/source/salt/salt/output/highstate.py", line 119, in <lambda>
                  key=lambda k: data[k].get('__run_num__', 0)):
              AttributeError: 'int' object has no attribute 'get'
     Started: 12:53:50.115281
    Duration: 219.549 ms
     Changes:   

Summary
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
```

**Comment by cachedout** [comment]
#19884 should clean this up for you. Thanks very much for letting us know about this!

**Comment by rallytime** [comment]
@amendlik It looks like this issue should be fixed in the impending 2015.2.0 release. Are you in a position where you could give the HEAD of 2015.2 branch a try to test this? If you want to wait for the release, that is fine, too. Just let us know if this is resolved for you when you get a moment to test again.

**Comment by amendlik** [comment]
It doesn't give me a stack trace anymore, but I think the output could look a lot better. Notice the color codes and warnings:
```
Total states run:     1
dev:
    ----------
    salt_|-test_|-test_|-state:
        ----------
        __run_num__:
            0
        changes:
            ----------
        comment:
            Run failed on minions: dev
            Failures:
                ?[0;31mdev:?[0m
                ?[0;31m----------?[0m
                    ?[0;31m      ID: failtest?[0m
                    ?[0;31mFunction: salt.function?[0m
                    ?[0;31m    Name: cmd.run_all?[0m
                    ?[0;31m  Result: False?[0m
                    ?[0;31m Comment: Running function cmd.run_all failed on minions: dev
                              Failures:
                                  dev:
                                      ----------
                                      pid:
                                          28892
                                      retcode:
                                          0
                                      stderr:
                                          
                                      stdout:
                                          Sat Feb 28 09:09:47 MST 2015?[0m
                    ?[0;31m Started: 09:09:46.414866?[0m
                    ?[0;31mDuration: 1108.13 ms?[0m
                ?[0;31m     Changes:   ?[0m
                   ?[1;31m Warnings: 'fun' and 'state' are invalid keyword arguments for
                              'salt.function'. If you were trying to pass additional data to be
                              used in a template context, please populate 'context' with 'key:
                              value' pairs. Your approach will work until Salt Carbon is out.
                              Please update your state files.?[0m
                ?[0;36m
                Summary
                ------------?[0m
                ?[0;32mSucceeded: 0?[0m
                ?[0;31mFailed:    1?[0m
                ?[1;31mWarnings:  1?[0m
                ?[0;36m------------
                Total states run:     1?[0m
            
        duration:
            2786.778 ms
        name:
            test
        result:
            False
        start_time:
            09:09:44.774251
        warnings:
            - 'fun' and 'state' are invalid keyword arguments for 'salt.state'. If you were trying to pass additional data to be used in a template context, please populate 'context' with 'key: value' pairs. Your approach will work until Salt Carbon is out. Please update your state files.
```

**Comment by rallytime** [comment]
@amendlik Thanks for the update. I think you might now be hitting #21120, at least for the warnings. I am not sure if the strange output formatting is related to that bug or not, but I suspect that it might be.

**Comment by s0undt3ch** [comment]
Once https://github.com/saltstack/salt/pull/21164 is merged, can you please try again?

**Comment by rallytime** [comment]
@amendlik Can you give this one more try with the referenced changes above? Try the HEAD of 2015.2 and let us know how it goes.

**Comment by amendlik** [comment]
I tested 6cf822561ac00ef38431547c357a3b0a1226437e. Now I get this exception on any function call from an orchestration:
```
Exception occurred in runner state.orch: Traceback (most recent call last):
  File "/source/salt/salt/client/mixins.py", line 327, in low
    data['return'] = self.functions[fun](*args, **kwargs)
  File "/source/salt/salt/runners/state.py", line 116, in orchestrate
    __jid_event__.fire_event({'data': ret, 'outputter': 'highstate'}, 'progress')
  File "/source/salt/salt/utils/event.py", line 530, in fire_event
    self.print_func(tag, data)
  File "/source/salt/salt/client/mixins.py", line 447, in print_async_event
    if isinstance(event['return'], dict) \
KeyError: 'return'
```

**Comment by cachedout** [comment]
@amendlik It looks like the fix didn't make its way into the correct branch. Could we trouble you to test this once more agains the HEAD of 2015.2? Thanks.

**Comment by amendlik** [comment]
It does return false now, which is the correct result. I still feel the output could be made clearer, such as indicating that the function failed because it does not exist.

```yaml
suffix:
    progress
dev:
    ----------
    salt_|-not.a.real.function_|-not.a.real.function_|-function:
        ----------
        __run_num__:
            0
        changes:
            ----------
            out:
                highstate
            ret:
                ----------
                dev:
                    'not' __virtual__ returned False
        comment:
            Running function not.a.real.function failed on minions: dev Function not.a.real.function ran on dev.
        duration:
            321.199
        name:
            not.a.real.function
        result:
            False
        start_time:
            19:47:07.563819
```

**Comment by cachedout** [comment]
@amendlik Could you please open a new issue with your thoughts about improving the output? I feel like we've solved the original issue here and would like to close this one.

**Comment by rallytime** [comment]
Since this original issue has been addressed, I am going to close this issue out. Feel free to open a new issue about improving the output as you see fit. Thanks again!
